src/Plugin/RulesAction/RulesComponentAction.php

```diff
@@ -0,0 +1,132 @@
+   *
+   * @var string[]
+   */
+  protected $saveLater = [];
+
+  /**
+   * Constructs an EntityCreate object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin ID for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Entity\EntityStorageInterface $storage
+   *   The entity storage service.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, EntityStorageInterface $storage) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->storage = $storage;
+    $this->componentId = $plugin_definition['component_id'];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity_type.manager')->getStorage('rules_component')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute() {
+    $rules_config = $this->storage->load($this->componentId);
+
+    // Setup an isolated execution state for this expression and pass on the
+    // necessary context.
+    $rules_component = $rules_config->getComponent();
+    foreach ($this->getContextValues() as $context_name => $context_value) {
+      $rules_component->setContextValue($context_name, $context_value);
+    }
+
+    // We don't use RulesComponent::execute() here since we don't want to
+    // auto-save immedialtely.
+    $state = $rules_component->getState();
+    $expression = $rules_component->getExpression();
+    $expression->executeWithState($state);
+
+    // Postpone auto-saving to the parent expression triggering this action.
+    foreach ($state->getAutoSaveSelectors() as $selector) {
+      $parts = explode('.', $selector);
+      $context_name = reset($parts);
+      if (array_key_exists($context_name, $this->context)) {
+        $this->saveLater[] = $context_name;
+      }
+      else {
+        // Otherwise we need to save here since it will not happen in the parent
+        // execution.
+        $typed_data = $state->fetchDataByPropertyPath($selector);
+        // The returned data can be NULL, only save it if we actually have
+        // something here.
+        if ($typed_data) {
+          // Things that can be saved must have a save() method, right?
+          // Saving is always done at the root of the typed data tree, for example
+          // on the entity level.
+          $typed_data->getRoot()->getValue()->save();
+        }
+      }
+    }
+
+    $provided_context = $rules_component->getProvidedContext();
+    foreach ($provided_context as $name) {
+      $this->setProvidedValue($name, $state->getVariableValue($name));
```

Comment:
Again, let's rather pass through the typed data object?
-> get the provided context object and set the variable via setContextData().